Only attempt to sync schemas that have some/any possible permissions

jeff303 opened the pull request:
Add new multimethod `have-any-schema-privileges?`, to the sync interface, that a driver can implement if it has a way to check for "any" schema-level permissions

Update `syncable-schemas` to call that method, if defined, in order to filter the output of `all-schemas`

Update Redshift driver to define implementation for `have-any-schema-privileges?` by calling the Redshift `HAS_SCHEMA_PRIVILEGE` function

jeff303 committed:
Only attempt to sync schemas that have some/any possible permissions

Add new multimethod have-any-schema-privileges?, to the sync interface, that a driver can implement if it has a way to check for "any" schema-level permissions

Update syncable-schemas to call that method, if defined, in order to filter the output of all-schemas

Update Redshift driver to define implementation for have-any-schema-privileges? by calling the Redshift HAS_SCHEMA_PRIVILEGE function

camsaul commented on modules/drivers/redshift/src/metabase/driver/redshift.clj:
```diff
@@ -251,3 +251,10 @@
        (qputil/default-query->remark query)))
+
+(defmethod sql-jdbc.sync/have-any-schema-privileges? :redshift [_ ^Connection conn ^String table-schema]
+  (with-open [stmt ^PreparedStatement (prepare-statement conn "SELECT HAS_SCHEMA_PRIVILEGE(?, 'USAGE');")]
+    (.setString stmt 1 table-schema)
+    (with-open [rs ^ResultSet (.executeQuery stmt)]
```
1.  Just add tag information to `prepare-statement` rather than having to tag it every time it's used.

   ```clj
   (defn- prepare-statement ^PreparedStatement [^Connection conn sql]
     (.prepareStatement conn sql
                        ResultSet/TYPE_FORWARD_ONLY
                        ResultSet/CONCUR_READ_ONLY
                        ResultSet/CLOSE_CURSORS_AT_COMMIT))
   ``` 

2.  (For future reference) generally you'd add the tag to the symbol e.g. `^PreparedStatement stmt (...)` instead of on the fn call form e.g. `stmt ^PreparedStatement (...)` -- although the latter does work. It looks unidiomatic tho

3. You don't need to tag Java methods like `.executeQuery`, Java is statically typed so it already knows `Statement.executeQuery` returns a `ResultSet` without you telling it

4.  You could combine this into one `with-open` statement if you did 

   ```clj
   (doto (prepare-statement ...) 
     (.setString 1 table-schema))
   ```

camsaul replied:
If you follow my suggestion and made a `remove-schemas-without-privileges` or `syncable-schemas` function you could optimize this a bit further. You can create the prepared statement just once and reuse it (just set a new string and fetch `with-open` a new result set for each schema). Since a big database might have a few thousand schemas we should try to optimize this a bit if we can

jeff303 replied:
This is done via a stateful transducer now in the Redshift version.

camsaul commented on modules/drivers/redshift/src/metabase/driver/redshift.clj:
```diff
@@ -251,3 +251,10 @@
        " */ "
        (qputil/default-query->remark query)))
+
+(defmethod sql-jdbc.sync/have-any-schema-privileges? :redshift [_ ^Connection conn ^String table-schema]
```
This isn't in the Clojure style guide but generally I think it looks nicer to have the args on a new line for `defmethod` forms -- that's the convention we've been using in most of the MB code

jeff303 replied:
No longer relevant with latest change.

camsaul commented on src/metabase/driver/sql_jdbc/sync/describe_database.clj:
You should just add the filter to the `eduction` above.

```clj
(defn- syncable-schemas
  [driver metadata]
  (eduction (comp
             (if (do-some-extra-filtering?)
               (filter my-filter-fn)
               identity)
             (remove (set (i/excluded-schemas driver))))
            (all-schemas metadata)))
```

It's a little weird to take `all-schemas`, realize the whole thing in memory as a vector, then create a lazy sequence where you filter out things based on the perms check function, then create an `Eduction` that transduces those lazy results with another filtering function. 

Something like

```clj
(defn- syncable-schemas
  [driver metadata]
  (eduction (comp
             (if (do-some-extra-filtering?)
               (filter my-filter-fn)
               identity)
             (remove (set (i/excluded-schemas driver))))
            (all-schemas metadata)))
```

should work.

camsaul replied:
You should also do the perms check *after* filtering by `i/excluded-schemas`, since those are hardcoded it will remove stuff without requiring a query. Do fast filtering first, then slow filtering second

camsaul replied:
There's a reason we were using `eduction` in the first place... Suppose someone with a mega database has 10k schemas (some places do one-schema-per-customer for example), but their user has privs for exactly one schema... the change you made now holds all 10k schema info map in memory at once *before* doing any filtering. If you use an eduction then you'd only ever need one schema info map in memory at once.

jeff303 replied:
I have updated the change to use transducers properly instead of realizing the entire sequence of all schemas.

camsaul commented on src/metabase/driver/sql_jdbc/sync/interface.clj:
```diff
@@ -27,6 +27,13 @@
   driver/dispatch-on-initialized-driver
   :hierarchy #'driver/hierarchy)
 
+(defmulti have-any-schema-privileges?
+  "Check if we have any sort of privileges for the given `schema`. If this method is defined for some driver, and it
+  returns false for the given schema, then we won't attempt to check for any tables within."
+  {:arglists '([driver ^java.sql.Connection connection ^String table-schema])}
+  driver/dispatch-on-initialized-driver
+  :hierarchy #'driver/hierarchy)
```
I think you should change this to something like `remove-schemas-without-privileges` so drivers that can determine them all at once can do so... checking 1000 schemas one-at-a-time is potentially a little slow, so it would be nice to give drivers the chance to optimize it a bit.

Better yet... why not just make `syncable-schemas` the new multimethod and let drivers filter out things however they please? Maybe they're filtering out ones based on some predefined blacklist, or maybe they're doing perms checking like Redshift does. The default impl can still use `excluded-schemas`

jeff303 replied:
OK, given everything outlined above, I think it makes the most sense to just have `syncable-schemas` as the multimethod, and duplicate a tiny bit in building up the `eduction` for Redshift.

camsaul reviewed (changes requested):
1.  Tag the function, not places that use it
2. We were using `eduction` for a reason -- please be careful not to write code that now causes everything to be in memory all at once. We don't want sync causing OOM issues again
3. I think you should generalize the new multimethod to something like `syncable-schemas`  so drivers can filter things out for whatever reason they wish, and can do so in an efficient way. For example, maybe Postgres can determine this info for 1000 schemas in a single query. We'd be crazy not to take advantage of that.  The current code isn't general enough to let us do those sort of optimizations

camsaul commented on src/metabase/driver/sql_jdbc/sync/interface.clj:
```diff
@@ -27,6 +27,13 @@
   driver/dispatch-on-initialized-driver
   :hierarchy #'driver/hierarchy)
 
+(defmulti have-any-schema-privileges?
```
One more note: you should add `^{:added "0.39.0"}` metadata on driver multimethods so driver authors can see when stuff was introduced more easily

jeff303 replied:
Done (for the new multi).

jeff303 committed:
PR feedback:

Fixing type annotations as per comments

Make syncable-schemas the multimethod, whose base implementation is the same as the previous private fn

Override syncable-schemas for Redshift to add a filtering step to the transducer pipeline to check for the schema privilege (implemented as a stateful transducer to reuse the PreparedStatement)

Making all-schemas public since it's now invoked from the Redshift driver

Fixing redshift-types-test to remove order flakiness

Fixing a few other places as per PR feedback

camsaul commented on modules/drivers/redshift/src/metabase/driver/redshift.clj:
```diff
@@ -251,3 +253,27 @@
        " */ "
        (qputil/default-query->remark query)))
+
+(defn- make-schema-usage-filter-xf [conn]
+  (fn [xf]
+    (let [stmt (prepare-statement conn "SELECT HAS_SCHEMA_PRIVILEGE(?, 'USAGE');")]
+      (fn
+        ([]
+         (xf))
+        ([result]
+         (.close stmt)
+         (xf result))
+        ([result table-schema]
+         (with-open [rs (.executeQuery (doto stmt (.setString 1 table-schema)))]
+           (let [has-perm? (and (.next rs)
+                                (.getBoolean rs 1))]
+             (if has-perm?
+               (xf result table-schema)
+               (do (log/tracef "Ignoring schema %s because no USAGE privilege on it" table-schema)
+                   result)))))))))
```
If any steps here throw an Exception it's going to leave the `PreparedStatement` open without closing it.

Sorry to keep making your life hard, but I think what you actually want to do here is reify `clojure.lang.IReduceInit` so you can use `with-open` with `prepare-statement`. Give me a second and I can show you what I mean

camsaul reviewed (changes requested):
Hey @jeff303 take a look at the tweaks I made to this PR in #15407, that's how I think you should approach it. Fell free to merge those changes into this PR if they make sense

Reifying an `IReduceInit` will let us have control over the lifecycle of the `PreparedStatement`. As it is, this PR would leak the `PreparedStatement` and never close it if there were any exceptions in the reducing code.

I also added a test for things since there wasn't one already. I wasn't sure how to actually test a schema with no permissions, so I mocked a few things and checked that it filters out schemas that don't exist instead, since by definition you won't have perms for something that doesn't exist.

camsaul committed:
Make syncable-schemas reducible (#15407)

jeff303 committed:
Add test that confirms that a real schemas with no permissions aren't synced

jeff303 commented:
> Hey @jeff303 take a look at the tweaks I made to this PR in #15407, that's how I think you should approach it. Fell free to merge those changes into this PR if they make sense
> 
> Reifying an `IReduceInit` will let us have control over the lifecycle of the `PreparedStatement`. As it is, this PR would leak the `PreparedStatement` and never close it if there were any exceptions in the reducing code.

Thanks for doing that.  I had a feeling there was a more correct way to do that.  Seems [this article](https://juxt.pro/blog/ontheflycollections-with-reducible) provides a good explanation.

> I also added a test for things since there wasn't one already. I wasn't sure how to actually test a schema with no permissions, so I mocked a few things and checked that it filters out schemas that don't exist instead, since by definition you won't have perms for something that doesn't exist.

I just added a test that confirms this with a real (temporarily created) schema and user.  I don't think we need to worry about the schema limit, anymore, since it appears to be something like [9900 now](https://docs.aws.amazon.com/redshift/latest/mgmt/amazon-redshift-limits.html) (instead of 256 as it apparently was in late 2015, as per comments in `session-schema-number`).

camsaul reviewed (approved):
cool. looks good to me. Thanks for figuring out how to add the test for schema permissions
